### PR TITLE
Support default/external json for open-mmlab models

### PR DESCRIPTION
**Pull request by xvjiarui**
This PR supported to specify open-mmlab models by json file. 
To override the default json file, first export `$MMCV_HOME=/your/path/`, then place a new json file under `$MMCV_HOME/urls`.

### CODE REVIEW

**Comment by hellock** [comment]
The proposed design:
- We provide a default json file `open-mmlab.json` in the mmcv package.
- By default the default json file is used for mapping urls.
- The default value of env `MMCV_HOME` is $HOME/.cache/mmcv
- If additional json files under `$MMCV_HOME` exists, we will load and merge them with the default one. The custom file may override some keys in the default one. The URI scheme is the filename, e.g., items in `local.json` can be obtained with `local://xxxx`.

File: docs/model_zoo.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+* `local://zzz`: The model filepaths provided in external json files.
+
+### Model URLs in JSON
+The json file consists of key-value pair of model name and its url/path.
+An example json file could be like:
+```json
+{
+    `model_a`: `https://example.com/models/model_a_9e5bac.pth,
+    `model_b`: `pretrain/model_b_ab3ef2c.pth`,
+}
```
Comment (hellock) [inline docs/model_zoo.md:17]:
```json
{
    "model_a": "https://example.com/models/model_a_9e5bac.pth",
    "model_b": "pretrain/model_b_ab3ef2c.pth",
}
```

File: docs/model_zoo.md [hunk 2]
```diff
@@ -0,0 +1,23 @@
+## Model Zoo
+
+The following types are supported for `filename` argument of `mmcv.load_checkpoint()`.
+* filepath: The filepath of the checkpoint.
+* link: The link to download the checkpoint. The `SHA256` prefix should be contained in the filename.
```
Comment (hellock) [inline docs/model_zoo.md:5]: - link -> `http://xxx` and `https://xxx`
- prefix -> postfix

File: mmcv/runner/checkpoint.py [hunk 3]
```diff
@@ -139,6 +122,23 @@ def get_torchvision_models():
...
+    mmcv_home = _get_mmcv_home()
+    default_json_path = osp.join(mmcv.__path__[0], 'urls/open_mmlab.json')
+    default_urls = load_file(default_json_path)
+    assert isinstance(default_urls, dict)
+    external_urls = dict()
+    for json_file in scandir(mmcv_home, suffix='.json'):
+        json_path = osp.join(mmcv_home, json_file)
+        urls = load_file(json_path)
+        assert set(external_urls.keys()).isdisjoint(set(urls.keys())), \
+            'External urls should not have overlapped keys'
```
Comment (hellock) [inline mmcv/runner/checkpoint.py:135]: We may print the overlapped keys in the error message

File: mmcv/runner/checkpoint.py [hunk 4]
```diff
@@ -139,6 +122,23 @@ def get_torchvision_models():
...
+        print(f'Update model urls with: {json_path}')
```
Comment (hellock) [inline mmcv/runner/checkpoint.py:136]: Is this printing necessary?

File: mmcv/runner/checkpoint.py [hunk 5]
```diff
@@ -162,8 +162,18 @@ def _load_checkpoint(filename, map_location=None):
         model_name = filename[14:]
         checkpoint = load_url_dist(model_urls[model_name])
     elif filename.startswith('open-mmlab://'):
+        model_urls = get_external_models()
         model_name = filename[13:]
-        checkpoint = load_url_dist(open_mmlab_model_urls[model_name])
+        checkpoint = load_url_dist(
+            model_urls[model_name], model_dir=_get_mmcv_home())
+    elif filename.startswith('local://'):
```
Comment (hellock) [inline mmcv/runner/checkpoint.py:169]: `local://` or `custom://`, which one is better?

File: docs/model_zoo.md [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+* `torchvison://xxx`: The model links in `torchvision.models`.Please refer to [torchvision](https://pytorch.org/docs/stable/torchvision/models.html) for details.
+* `open-mmlab://yyy`: The model links provided in default and external json files.
+* `local://zzz`: The model filepaths provided in external json files.
```
Comment (hellock) [inline docs/model_zoo.md:8]: The two `external json files` are unclear.

File: mmcv/runner/checkpoint.py (hunk 4 above)
Comment (xvjiarui) [reply mmcv/runner/checkpoint.py:136]: We may need to inform users about non default model links/files.

File: mmcv/runner/checkpoint.py [hunk 7]
```diff
@@ -139,11 +122,27 @@ def get_torchvision_models():
     return model_urls
 
 
+def get_external_models():
+    mmcv_home = _get_mmcv_home()
+    default_json_path = osp.join(mmcv.__path__[0], 'model_zoo/open_mmlab.json')
```
Comment (hellock) [inline mmcv/runner/checkpoint.py:127]: Check if the json file is included in the distribution.

File: mmcv/runner/checkpoint.py [hunk 8]
```diff
@@ -162,8 +161,17 @@ def _load_checkpoint(filename, map_location=None):
...
-        checkpoint = load_url_dist(open_mmlab_model_urls[model_name])
+        model_url = model_urls[model_name]
+        # check if is url
+        if model_url.startswith(('http://', 'https://')):
+            checkpoint = load_url_dist(model_url)
+        else:
+            filename = osp.join(_get_mmcv_home(), model_url)
+            if not osp.isfile(filename):
+                raise IOError(f'{filename} is not a checkpoint file')
+            checkpoint = torch.load(filename, map_location=map_location)
```
Comment (hellock) [inline mmcv/runner/checkpoint.py:174]: We may consider package the above logic into a method, to support load checkpoint from a url or local file, (and potentially s3).

**Pull request merged**